Refactor for tests: start work on Emitter.

m-kostrzewa opened the pull request:
In order to write tests, we could implement a test version of emit() method and assert like

```
fn it_checks_imports() {
   // ...
  let mut fake_emitter: ...
  run("test_cases/invalid_something/", fake_emitter)
  assert_contains(fake_emitter.emitted_items, "Error: Invalid something in line 10")
}
```

In order to do that we need to be able to inject some kind of FakeEmitter that has its own emit() method. It's because the "old" emitter just prints everything to stdout, so we can't easily create assertions.

Changes:
- main.rs and lib.rs
  - Move implementation to lib.rs and main to main.rs (so that our tests can test the lib.rs).
  - Also, update Cargo.toml so that both lib and ctl tool are compiled
- Refactors
  - Create Emitter struct with emit() method
  - Create helper struct EmittedItem
  - Pass Emitter to all functions that used emit()
  - Change some field types to store everything by value
  - Move message formating to emit_message() function in main.rs

This code needs polishing and can be refactored further, but I wanted to see your comments before continuing. Especially since I'm a Rust noob.

m-kostrzewa committed:
Refactor for tests: start work on Emitter.

krdln commented:
Thanks for the PR, overally I like the idea of making it more testable. (I have some nitpicks, though)

One thing I'm not sure about is that we're no longer emit messages online. We could make `Emitter` a trait instead, with `RealEmitter` and `TestEmitter` instead. What do you think? (This way we also won't have to iterate over the vec in `main`).

Regarding testing, I think that in addition to "full run" tests as you described, we could also test the `parse` and `analyze` method separately. (That would require splitting `parse` to `parse_file` and `parse_string`)

krdln committed:
Merge branch 'master' into emitter-dip

krdln committed:
Get rid of warnings

krdln committed:
Rename the binary back to shelly

Cargo supports library + binary setup by default,
no need to specify it in config.

krdln reviewed (approved):
I'm going to merge it as is, as I plan some refactors after which it'd be hard to keep the branch up to date.
